x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -237,4 +392,293 @@ public void onFailure(Exception e) {
+                    if (searchAfter != null) {
+                        searchSource.searchAfter(searchAfter);
+                    }
+                    clientWithOrigin.execute(SearchAction.INSTANCE, searchRequest, new ActionListener<>() {
+                        @Override
+                        public void onResponse(SearchResponse response) {
+                            maintenanceTask.total.compareAndSet(0L, response.getHits().getTotalHits().value);
+                            maintenanceTask.searchResponse = response;
+                            maintenanceTask.searchAfter = null;
+                            executeNext(maintenanceTask);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            complete(maintenanceTask, e);
+                        }
+                    });
+                    return;
+                }
+
+                final SearchHit[] searchHits = searchResponse.getHits().getHits();
+                if (searchHits != null && searchHits.length > 0) {
+                    final ClusterState state = clusterService.state();
+                    final BulkRequest bulkRequest = new BulkRequest();
+
+                    final TimeValue retention = periodicTaskRetention;
+                    final Instant expirationTime = Instant.ofEpochMilli(threadPool.absoluteTimeInMillis())
+                        .minus(retention.duration(), retention.timeUnit().toChronoUnit());
+
+                    // compute the list of existing searchable snapshots once
+                    Map<String, Set<String>> knownSnapshots = existingSnapshots;
+                    if (knownSnapshots == null) {
+                        knownSnapshots = listSearchableSnapshots(state);
+                        existingSnapshots = knownSnapshots;
+                    }
+
+                    final Set<String> knownRepositories = state.metadata()
+                        .custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY)
+                        .repositories()
+                        .stream()
+                        .map(RepositoryMetadata::name)
+                        .collect(Collectors.toSet());
+
+                    Object[] lastSortValues = null;
+                    for (SearchHit searchHit : searchHits) {
+                        lastSortValues = searchHit.getSortValues();
+                        assert searchHit.getId() != null;
+                        assert searchHit.hasSource();
+
+                        try {
+                            final CachedBlob cachedBlob = CachedBlob.fromSource(searchHit.getSourceAsMap());
+                            if (cachedBlob.creationTime().isAfter(expirationTime)) {
+                                logger.trace("blob store cache entry with id [{}] was created recently, skipping", searchHit.getId());
+                                continue;
+                            }
```

Comment:
nit: I think this should move to after we detected that this is a snapshot to delete and then let the log message go to debug instead.

Reply:
Agreed -> 1966a839313560d0d624528417e131c09072e6eb